File: sercop-unified/frontend/src/services/swiftSectionConfigService.ts
import { API_BASE_URL } from '../config/api.config';

/**
 * Configuración de una sección SWIFT
 */
export interface SwiftSectionConfig {
  id: string;
  sectionCode: string;
  labelKey: string;
  descriptionKey?: string;
  messageType: string;
  displayOrder: number;
  icon?: string;
  isActive: boolean;
}

const BASE_URL = `${API_BASE_URL}/swift-section-config`;

/**
 * Servicio para obtener configuraciones de secciones SWIFT desde el backend
 */
export const swiftSectionConfigService = {
  /**
   * Obtiene las secciones activas para un tipo de mensaje
   */
  async getSectionsByMessageType(messageType: string, includeInactive = false): Promise<SwiftSectionConfig[]> {
    const url = `${BASE_URL}/${messageType}?includeInactive=${includeInactive}`;
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(`Error fetching sections for ${messageType}: ${response.statusText}`);
    }

    return response.json();
  },

  /**
   * Obtiene los tipos de mensaje que tienen secciones configuradas
   */
  async getConfiguredMessageTypes(): Promise<string[]> {
    const url = `${BASE_URL}/message-types`;
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(`Error fetching message types: ${response.statusText}`);
    }

    return response.json();
  },

  /**
   * Obtiene una sección específica
   */
  async getSectionByCode(messageType: string, sectionCode: string): Promise<SwiftSectionConfig | null> {
    const url = `${BASE_URL}/${messageType}/${sectionCode}`;
    const response = await fetch(url);

    if (response.status === 404) {
      return null;
    }

    if (!response.ok) {
      throw new Error(`Error fetching section: ${response.statusText}`);
    }

    return response.json();
  }
};

export default swiftSectionConfigService;

File: sercop-unified/backend-java/src/main/java/com/globalcmx/api/service/SwiftSectionConfigService.java
package com.globalcmx.api.service;

import com.globalcmx.api.dto.swift.SwiftSectionConfigDTO;
import com.globalcmx.api.readmodel.entity.SwiftSectionConfig;
import com.globalcmx.api.readmodel.repository.SwiftSectionConfigRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Servicio para gestionar configuraciones de secciones SWIFT
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class SwiftSectionConfigService {

    private final SwiftSectionConfigRepository repository;

    /**
     * Obtiene todas las secciones activas para un tipo de mensaje
     */
    @Transactional(readOnly = true)
    public List<SwiftSectionConfigDTO> getActiveSectionsByMessageType(String messageType) {
        log.debug("Fetching active sections for message type: {}", messageType);
        return repository.findActiveByMessageTypeOrdered(messageType)
                .stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    /**
     * Obtiene todas las secciones para un tipo de mensaje (incluye inactivas)
     */
    @Transactional(readOnly = true)
    public List<SwiftSectionConfigDTO> getAllSectionsByMessageType(String messageType) {
        log.debug("Fetching all sections for message type: {}", messageType);
        return repository.findByMessageTypeOrderByDisplayOrder(messageType)
                .stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    /**
     * Obtiene una sección específica por código y tipo de mensaje
     */
    @Transactional(readOnly = true)
    public Optional<SwiftSectionConfigDTO> getSectionByCodeAndMessageType(String sectionCode, String messageType) {
        return repository.findBySectionCodeAndMessageType(sectionCode, messageType)
                .map(this::toDTO);
    }

    /**
     * Obtiene todos los tipos de mensaje que tienen secciones configuradas
     */
    @Transactional(readOnly = true)
    public List<String> getConfiguredMessageTypes() {
        return repository.findDistinctMessageTypes();
    }

    /**
     * Convierte entidad a DTO
     */
    private SwiftSectionConfigDTO toDTO(SwiftSectionConfig entity) {
        return SwiftSectionConfigDTO.builder()
                .id(entity.getId())
                .sectionCode(entity.getSectionCode())
                .labelKey(entity.getLabelKey())
                .descriptionKey(entity.getDescriptionKey())
                .messageType(entity.getMessageType())
                .displayOrder(entity.getDisplayOrder())
                .icon(entity.getIcon())
                .isActive(entity.getIsActive())
                .build();
    }
}
